broker.go

```diff
@@ -56,20 +62,24 @@ func (br *defaultBroker) Publish(pub Sender, msg *Publish) {
 
 // Subscribe subscribes the client to the given topic.
 func (br *defaultBroker) Subscribe(sub Sender, msg *Subscribe) {
+	id := NewID()
+
+	br.lock.Lock()
 	if _, ok := br.routes[msg.Topic]; !ok {
 		br.routes[msg.Topic] = make(map[ID]Sender)
 	}
-	id := NewID()
 	br.routes[msg.Topic][id] = sub
-
 	br.subscriptions[id] = msg.Topic
+	br.lock.Unlock()
 
 	sub.Send(&Subscribed{Request: msg.Request, Subscription: id})
 }
 
 func (br *defaultBroker) Unsubscribe(sub Sender, msg *Unsubscribe) {
+	br.lock.Lock()
 	topic, ok := br.subscriptions[msg.Subscription]
 	if !ok {
+		br.lock.Unlock()
```

Comment:
never mind, I see you're trying to not lock the Send. You might want to use a scoping block to make defer work just for this scope.